src/uniform.jl

```diff
@@ -0,0 +1,34 @@
+"""
+    UniformQuantization(n::Int)
+
+Quantize colors in RGB color space by dividing each dimension of the ``[0, 1]³``
+RGB color cube into `n` equidistant steps for a total of `n³` cubes of equal size.
+Each color in `cs` is then quantized to the center of the cube it is in.
+Only unique colors are returned. The amount of output colors is therefore bounded by `n³`.
+"""
+struct UniformQuantization <: AbstractColorQuantizer
+    n::Int
+
+    function UniformQuantization(n)
+        return n < 1 ? throw(ArgumentError("n has to be ≥ 1, got $(n).")) : new(n)
+    end
+end
+
+const UQ_FLOATTYPE = Float32
+const UQ_COLORTYPE = RGB{UQ_FLOATTYPE}
+const UQ_HALF = UQ_FLOATTYPE(0.5)
+
+(alg::UniformQuantization)(cs) = alg(convert.(UQ_COLORTYPE, cs))
+function (alg::UniformQuantization)(cs::AbstractArray{UQ_COLORTYPE})
+    n = UQ_FLOATTYPE(alg.n)
+    return unique(map(c -> _uniform_quantization(c, n), cs))
+end
+
+function _uniform_quantization(c::UQ_COLORTYPE, n::UQ_FLOATTYPE)
+    return mapc(x -> _uniform_quantization(x, n), c)
+end
+function _uniform_quantization(x::UQ_FLOATTYPE, n::UQ_FLOATTYPE)
+    x ≤ 0 && return 1 / (2 * n)
+    x ≥ 1 && return (2 * n - 1) / (2 * n)
+    return (round(x * n - UQ_HALF) + UQ_HALF) / n
+end
```

Comment:
This is okay. But I get something faster by making it a plain lookup table (The key idea is to pass `T,N` to the compiler so that we can do the computation in the compilation stage using [generated function](https://docs.julialang.org/en/v1/manual/metaprogramming/#Generated-functions)):

```julia
using ImageCore, BenchmarkTools

struct UniformQuantization{T, N} end

@inline function quantize(::UniformQuantization{T,N}, x::T) where {T,N}
    _naive_quantize(UniformQuantization{T,N}(), x)
end

@inline function quantize(::UniformQuantization{T,N}, x::T, n) where {T<:Union{N0f8,},N}
    @inbounds _build_lookup_table(UniformQuantization{T,N}())[reinterpret(FixedPointNumbers.rawtype(T), x) + 1]
end

@generated function _build_lookup_table(::UniformQuantization{T,N}) where {T<:FixedPoint,N}
    # This errors when T requires more than 32 bits
    tmax = typemax(FixedPointNumbers.rawtype(T))
    table = Vector{T}(undef, tmax + 1)
    for raw_x in zero(tmax):tmax
        x = reinterpret(T, raw_x)
        table[raw_x + 1] = _naive_quantize(UniformQuantization{T,N}(), x)
    end
    return table
end

function _naive_quantize(::UniformQuantization{T,N}, x::T) where {T,N}
    x ≤ 0 && return 1 / (2 * N)
    x ≥ 1 && return (2 * N - 1) / (2 * N)
    return (round(x * N - T(0.5)) + T(0.5)) / N
end
```

and I get

```julia
X = rand(N0f8, 512, 512)
Q = UniformQuantization{N0f8,32}()
@btime quantize.(Ref(Q), $X); #   95.689 μs (7 allocations: 256.24 KiB)
```

In the meantime, the current version is:

```julia
X = rand(N0f8, 512, 512)
Q = UniformQuantization(32)
@btime Q(X); #   5.336 ms (17 allocations: 6.01 MiB)
```

Reply:
I now have both in the code for maximum flexibility w.r.t. input data types. However the benchmarks are a bit more complicated:

```Julia
alg = UniformQuantization(32)
for T in (N0f8, Float32, RGB{N0f8}, RGB{Float32})
    println("Timing $T:")
    img = rand(T, 512, 512)
    @btime alg($img)
    @btime alg.($img) # skips `unique`
end
```
```
Timing N0f8:
  25.436 ms (15 allocations: 1.50 MiB)
  91.459 μs (4 allocations: 256.17 KiB)
Timing Float32:
  4.659 ms (17 allocations: 6.01 MiB)
  75.006 μs (3 allocations: 1.00 MiB)
Timing RGB{N0f8}:
  33.910 ms (38 allocations: 1.38 MiB)
  270.763 μs (3 allocations: 768.06 KiB)
Timing RGB{Float32}:
  7.663 ms (39 allocations: 5.28 MiB)
  344.923 μs (3 allocations: 3.00 MiB)
  ```
  And building the look-up table for `N0f32` freezes my Julia session.

Reply:
We certainly cannot make it a generic implementation here -- too risky here whether we can delegate so much computation to compiler. We only need those really matters in practice.
For generic types, a runtime lookup table (using Dict{Type,Vector} ) can be used instead.